### PR TITLE
add more flow annotations to RippleApi

### PR DESCRIPTION
**Pull request by darkdarkdragon**

### CODE REVIEW

File: src/api/common/utils.js [hunk 1]
```diff
@@ -17,7 +17,7 @@ function xrpToDrops(xrp: string): string {
   return (new BigNumber(xrp)).times(1000000.0).floor().toString();
 }
 
-function toRippledAmount(amount: Amount): string|Amount {
+function toRippledAmount(amount: AmountRippled): string|AmountRippled {
```
Comment (sublimator) [inline src/api/common/utils.js]: This doesn't seem right. It's taking an `AmountRippled` as input ? I believe this function takes a RippleAPI amount and converts it to the json form as spoken by rippled.

Comment (darkdarkdragon) [inline src/api/common/utils.js]: You're right. My mistake.

**Comment by darkdarkdragon** [comment]
this is errors in `server-test.js` (not related to this pr)

File: src/api/ledger/pathfind.js [hunk 2]
```diff
@@ -77,13 +71,15 @@ function addDirectXrpPath(paths, xrpBalance) {
...
-function isRippledIOUAmount(amount) {
+function isRippledIOUAmount(amount: Amount) {
   // rippled XRP amounts are specified as decimal strings
   return (typeof amount === 'object') &&
     amount.currency && (amount.currency !== 'XRP');
 }
 
-function conditionallyAddDirectXRPPath(remote, address, paths, callback) {
+function conditionallyAddDirectXRPPath(remote: Remote, address: string,
+  paths, callback
```
Comment (clark800) [inline src/api/ledger/pathfind.js]: Is there supposed to be an annotation on `paths`?

File: src/api/common/types.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
+/* @flow */
+'use strict';
+
+export type AmountRippled = {
```
Comment (clark800) [inline src/api/common/types.js]: Could we call it `RippledAmount`?

File: src/api/common/types.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+// Amount where counterparty and value are optional
+export type LaxLaxAmount = {
+  currency: string,
+  value?: string,
+  counterparty?: string
+}
+
+// A currency-counterparty pair, or just currency if it's XRP
+export type Issue = LaxLaxAmount;
```
Comment (clark800) [inline src/api/common/types.js]: An `Issue` shouldn't allow `value`

File: src/api/ledger/utils.js [hunk 5]
```diff
@@ -48,21 +58,20 @@ function getRecursive(getter: Getter, limit?: number, callback: Callback) {
...
   const issuer = amount.counterparty === undefined ?
-    amount.issuer : amount.counterparty;
+    (amount.issuer !== undefined ? amount.issuer : undefined) :
+    amount.counterparty;
   const withIssuer = _.assign({}, amount, {issuer: issuer});
   return _.omit(withIssuer, 'counterparty');
 }
 
-type Order = {taker_gets: Amount, taker_pays: Amount}
+type BOrder = {taker_gets: Issue, taker_pays: Issue}
```
Comment (clark800) [inline src/api/ledger/utils.js]: I think we should avoid abbreviations. I don't know what "B" means.

File: src/api/transaction/payment.js [hunk 6]
```diff
@@ -6,19 +6,45 @@ const validate = utils.common.validate;
 const toRippledAmount = utils.common.toRippledAmount;
 const Transaction = utils.common.core.Transaction;
 const ValidationError = utils.common.errors.ValidationError;
+// import type {Transaction} from '../../core/transaction';
```
Comment (clark800) [inline src/api/transaction/payment.js]: Don't need this anymore?

**Comment by darkdarkdragon** [comment]
@clark800 BTW, [here](https://github.com/darkdarkdragon/ripple-lib/blob/develop-flow-annotate-4/src/api/ledger/pathfind.js#L24), `destination_amount` that is returned from server gets overwritten with value from the function param (from the user). Is that ok?

File: src/api/transaction/utils.js [hunk 7]
```diff
@@ -54,11 +58,13 @@ function prepareTransaction(transaction: any, remote: any, instructions: any,
 
   function prepareMaxLedgerVersion(callback_) {
     if (instructions.maxLedgerVersion !== undefined) {
-      txJSON.LastLedgerSequence = parseInt(instructions.maxLedgerVersion, 10);
+      txJSON.LastLedgerSequence =
+        typeof instructions.maxLedgerVersion === undefined ?
```
Comment (clark800) [inline src/api/transaction/utils.js]: This is already in a branch where `instructions.maxLedgerVersion !== undefined`

**Comment by clark800** [comment]
I think that's because the `ripple_path_find` command doesn't return the `destination_amount`, though `path_find` does. We used to use `ripple_path_find`, but now we're using `path_find`. I think the user and response values should be the same, but maybe the best option would be to add `destination_amount` only if it is not already in the response.

File: src/api/transaction/utils.js (hunk 7 above)
Comment (darkdarkdragon) [inline src/api/transaction/utils.js]: I added this to preserve old behavior - `parseInt` returns `NaN` if value isn't something like digit.
If it is not needed, I can just do assign here.

File: src/api/common/types.js (hunk 3 above)
Comment (darkdarkdragon) [inline src/api/common/types.js]: maybe better to rename it to `RippledAmountIOU` and rename `SourceAmount` to `RippledAmount`

File: src/api/transaction/utils.js (hunk 7 above)
Comment (clark800) [inline src/api/transaction/utils.js]: I don't think it's possible, `typeof` always returns a string.

File: src/api/common/types.js (hunk 3 above)
Comment (clark800) [inline src/api/common/types.js]: That sounds good

File: src/api/transaction/utils.js (hunk 7 above)
Comment (darkdarkdragon) [inline src/api/transaction/utils.js]: My error, it intended to be `typeof instructions.maxLedgerVersion === 'undefined'`
strange that flow didn't catch this :disappointed:

Comment (clark800) [inline src/api/transaction/utils.js]: But that equality is guaranteed to always be false

Comment (darkdarkdragon) [inline src/api/transaction/utils.js]: Yeah, it is me being stupid :disappointed:

**Comment by darkdarkdragon** [comment]
@clark800 done

**Comment by clark800** [comment]
LGTM

**Pull request merged**